overlord/hookstate/ctlcmd/helpers.go

```diff
@@ -171,6 +181,84 @@ func runServiceCommand(context *hookstate.Context, inst *servicestate.Instructio
 	}
 }
 
+func validateSnapAndCompsNames(names []string, ctxSnap string) ([]string, error) {
+	comps := make([]string, len(names))
+	for i, name := range names {
+		snap, comp := snap.SplitSnapComponentInstanceName(name)
+		// if snap is present it must be the context snap for the moment
+		if snap != "" && snap != ctxSnap {
+			return nil, fmt.Errorf("cannot install snaps using snapctl yet")
+		}
+		if err := naming.ValidateSnap(comp); err != nil {
+			return nil, err
+		}
+		comps[i] = comp
+	}
+	return comps, nil
+}
+
+type managementCommandTp int
+
+const (
+	installManagementCommand managementCommandTp = iota
+	removeManagementCommand
+)
+
+type managementCommand struct {
+	typ        managementCommandTp
+	components []string
+}
+
+func runSnapManagementCommand(context *hookstate.Context, cmd *managementCommand) error {
+	st := context.State()
+	var tts []*state.TaskSet
+	var err error
+	var cmdStr, cmdVerb string
+
+	st.Lock()
+	switch cmd.typ {
+	case installManagementCommand:
+		tts, err = snapstateInstallComponents(st, cmd.components)
+		cmdStr = "install"
+		cmdVerb = "Installing"
+	case removeManagementCommand:
+		tts, err = snapstateRemoveComponents(st, cmd.components)
+		cmdStr = "remove"
+		cmdVerb = "Removing"
+	default:
+		err = fmt.Errorf("internal error: %q is not a valid snap management command", cmd.typ)
+	}
+	st.Unlock()
+	if err != nil {
+		return err
+	}
+
+	if !context.IsEphemeral() {
+		// Differently to service control commands, we always queue the
+		// management tasks if run from a hook.
+		return queueCommand(context, tts)
+	}
+
+	st.Lock()
+	chg := st.NewChange("snapctl-"+cmdStr,
+		fmt.Sprintf("%s components %v for snap %s",
+			cmdVerb, cmd.components, context.InstanceName()))
+	for _, ts := range tts {
+		chg.AddAll(ts)
+	}
+	st.EnsureBefore(0)
+	st.Unlock()
+
+	select {
+	case <-chg.Ready():
+		st.Lock()
+		defer st.Unlock()
+		return chg.Err()
+	case <-time.After(configstate.ConfigureHookTimeout()):
```

Comment:
this is the ephemeral case though so we are outside of hooks, we might want to decide on a different, much longer timeout?

Reply:
I've changed to 10 minutes now, should be enough.